source/client/service_impl.cc

```diff
@@ -0,0 +1,94 @@
+#include "client/service_impl.h"
+
+#include <grpc++/grpc++.h>
+
+#include "client/client.h"
+#include "client/options_impl.h"
+
+namespace Nighthawk {
+namespace Client {
+
+void ServiceImpl::handleExecutionRequest(const nighthawk::client::ExecutionRequest& request) {
+  ASSERT(running_, "running_ ought to be set");
+  nighthawk::client::ExecutionResponse response;
+  response.mutable_error_detail()->set_code(grpc::StatusCode::INTERNAL);
+
+  OptionsPtr options;
+  try {
+    options = std::make_unique<OptionsImpl>(request.start_request().options());
+  } catch (Envoy::EnvoyException exception) {
+    response.mutable_error_detail()->set_message(exception.what());
+    writeResponseAndFinish(response);
+    return;
+  }
+
+  // We scope here because the ProcessImpl instance must be destructed before we write the response
+  // and set running to false.
+  {
+    ProcessImpl process(*options, time_system_);
+    OutputCollectorFactoryImpl output_format_factory(time_system_, *options);
+    auto logging_context = std::make_unique<Envoy::Logger::Context>(
+        spdlog::level::from_str(options->verbosity()), "[%T.%f][%t][%L] %v", log_lock_);
+    auto formatter = output_format_factory.create();
+    if (process.run(*formatter)) {
+      response.clear_error_detail();
+      *(response.mutable_output()) = formatter->toProto();
+    } else {
+      response.mutable_error_detail()->set_message("Unknown failure");
+    }
+  }
+  writeResponseAndFinish(response);
+}
+
+void ServiceImpl::writeResponseAndFinish(const nighthawk::client::ExecutionResponse& response) {
+  if (!stream_->Write(response)) {
+    ENVOY_LOG(warn, "Stream write failed");
+  }
+  running_ = false;
+}
+
+void ServiceImpl::waitForRunnerThreadCompletion() {
+  if (runner_thread_.joinable()) {
```

Comment:
I had one thought on a good pattern here. What if instead of an explicit thread, we use `std::async` and `std::future`? Could we avoid a shared `running_` flag with this pattern?